src/coreclr/jit/compiler.hpp

```diff
@@ -4970,6 +4970,61 @@ unsigned Compiler::fgRunDfs(VisitPreorder visitPreorder, VisitPostorder visitPos
     return preOrderIndex;
 }
 
+//------------------------------------------------------------------------
+// fgVisitBlocksInLoopAwareRPO: Visit the blocks in 'dfsTree' in reverse post-order,
+// but ensure loop bodies are visited before loop successors.
+//
+// Type parameters:
+//   TFunc - Callback functor type
+//
+// Parameters:
+//   dfsTree     - The DFS tree of the flow graph
+//   blockToLoop - A mapping of each block in 'dfsTree' to its containing loop, if applicable
+//   func        - Callback functor that operates on a BasicBlock*
+//
+// Returns:
+//   A postorder traversal with compact loop bodies.
+//
+template <typename TFunc>
+void Compiler::fgVisitBlocksInLoopAwareRPO(FlowGraphDfsTree* dfsTree, BlockToNaturalLoopMap* blockToLoop, TFunc func)
+{
+    assert(dfsTree != nullptr);
+    assert(blockToLoop != nullptr);
+
+    EnsureBasicBlockEpoch();
+    BlockSet visitedBlocks(BlockSetOps::MakeEmpty(this));
```

Comment:
It would be better to use the post order number that you can get from the DFS tree.

Reply:
Good point, I should probably make a note of phasing this `bbNum` dependency out elsewhere.